### PR TITLE
Resolve the atten weight making NaN in MultiheadAttention forward func

### LINKED ISSUE #24816: Transformer Encoder Layer with src_key_padding makes NaN
I'm using PyTorch 1.3 and this is sample code for my example.

```python
import torch
import torch.nn as nn

enc = nn.TransformerEncoderLayer(3,1)
model = nn.TransformerEncoder(enc, 6)

x = torch.Tensor([[[1,2,3],[0,5,6]],[[-1,-2,0],[-4,-5,0]],[[1,2,3],[0,5,6]],[[-1,-2,0],[-4,-5,0]]])
y = torch.tensor([[0,0],[1,0],[1,1],[1,1]]).bool()
print(x.size())
print(y.size())
print(y)
output = model(src=x.transpose(0,1), src_key_padding_mask=y).transpose(0,1)
print(output)

>> torch.Size([4, 2, 3])
torch.Size([4, 2])
tensor([[False, False],
        [ True, False],
        [ True,  True],
        [ True,  True]])
tensor([[[-0.0933, -1.1754,  1.2687],
         [-0.4194, -0.9599,  1.3794]],

        [[ 0.9560, -1.3805,  0.4246],
         [ 0.9844, -1.3715,  0.3871]],

        [[    nan,     nan,     nan],
         [    nan,     nan,     nan]],

        [[    nan,     nan,     nan],
         [    nan,     nan,     nan]]], grad_fn=<TransposeBackward0>)
```

If I have all padded sequence with padding mask, this makes NaN output.
and backward() of this makes every NaN of outputs for next step of iterations.

I tried 
```python
output[torch.isnan(output)] = 0 
```
but it wasn't helpful, every output became zero. (mean still NaN for every output value)

How can I solve this?

cc @SsnL

### PR DESCRIPTION
**Pull request by ZhuBaohe**
This PR resolves [#24816](https://github.com/pytorch/pytorch/issues/24816#issue-481976681).

### CODE REVIEW

**Comment by zhangguanheng66** [comment]
I don't doubt the benefits to move the norm layer to the front end. However, there're some debates about the order and we have realized the discussions since the very beginning. If we accept the changes in this PR, there will be another request to have the norm layer on the back end. And it will just keep unsettled.

Our strategy is to keep consistent with the original paper (as we cite in the docs). If you add a norm layer between the embedding and the transformer, it will be effectively the same as you proposed here.

Regarding to the successive norm layers in the TransformerEncoder/TransformerDecoder, it will cover the cases when there is no norm layer on the back end of a custom TransformerEncoderLayer (an example is the layout proposed in this PR). With or without a norm layer in the TransformerEncoder won't change the results of the current version of `nn.Transformer` module. IMO, a norm layer on the back end of `nn.TransformerEncoder` gives some flexibility for some other custom transformer encoder/decoder layers.

**Comment by zhangguanheng66** [comment]
It's not a problem in `nn.MultiheadAttention` or `nn.Transformer`.

For the special case in the original issue, the user fully pads a batch so a tensor only filled with the `-inf` values is passed to `nn.softmax` function.  `nn.softmax` has to return a vector of `nan`. Will check with the team if this behavior is expected.

**Review by zhangguanheng66 — COMMENTED** [review]
It's not a problem in `nn.MultiheadAttention` or `nn.Transformer`.

For the special case in the original issue, the user fully pads a batch. As a result, within `nn.MultiheadAttention`, a tensor only filled with the `-inf` values is passed to `nn.softmax` function. `nn.softmax` has to return a vector of `nan`. Will check with the team if this behavior is expected.

Check the following code:
```
import torch
x = torch.Tensor([[[float("-inf"), float("-inf"), float("-inf")]]])
softmax = torch.nn.Softmax(dim=-1)
softmax(x)

>>> tensor([[[   nan,    nan,    nan]]])
```
